### PR TITLE
Prevent app code from blocking the connection loop 

### PR DESCRIPTION
**Pull request by Tratcher**
#2657 Each connection is multiplexed and has a loop that reads and processes frames. That processing must not ever get blocked by application code.

I fixed one code path where the initial request is dispatched to the app synchronously and the connection loop may get blocked by app code.

The only other case I found is for RST frames. https://github.com/aspnet/KestrelHttpServer/issues/2462#issuecomment-397702247. RequestAborted is dispatched on a background thread but that can be affected by ApplicationSchedulingMode. What's the guidance for ApplicationSchedulingMode? Should Inline mode assume the developer is going to write ideal code and never block the app loop? Or are there some conditions where we don't trust the app enough and always offload to a background thread like with the initial request above?

### CODE REVIEW

**Comment by halter73** [comment]
Looks like this is causing Http2ConnectionTests.ConnectionResetLoggedWithActiveStreams to fail

**Comment by halter73** [comment]
> What's the guidance for ApplicationSchedulingMode? Should Inline mode assume the developer is going to write ideal code and never block the app loop?

Yes.

File: src/Kestrel.Core/Internal/Http/HttpProtocol.cs [hunk 1]
```diff
@@ -1343,7 +1343,7 @@ private Pipe CreateRequestBodyPipe()
             (
                 pool: _context.MemoryPool,
                 readerScheduler: ServiceContext.Scheduler,
-                writerScheduler: PipeScheduler.Inline,
+                writerScheduler: PipeScheduler.ThreadPool,
```
Comment (halter73) [inline src/Kestrel.Core/Internal/Http/HttpProtocol.cs]: This being inline is fine for HTTP/1.1 where there's no multiplexing. We should make this virtual, and only change the HTTP/2 behavior.

Comment (halter73) [reply src/Kestrel.Core/Internal/Http/HttpProtocol.cs]: Also this affects the scheduling of server code, not app code, so I don't see how this could affect whether or not the app code could block the connection loop.

File: src/Kestrel.Core/Internal/Http2/Http2Connection.cs [hunk 2]
```diff
@@ -672,7 +672,15 @@ private void StartStream<TContext>(IHttpApplication<TContext> application)
...
-            _ = _currentHeadersStream.ProcessRequestsAsync(application);
+            // Must not allow app code to block the connection handling loop.
+            System.Threading.ThreadPool.UnsafeQueueUserWorkItem(state =>
+                {
+                    var tuple = (Tuple<IHttpApplication<TContext>, Http2Stream>)state;
+                    var app = tuple.Item1;
+                    var currentStream = tuple.Item2;
+                    _ = currentStream.ProcessRequestsAsync(app);
+                },
+                new Tuple<IHttpApplication<TContext>,Http2Stream>(application, _currentHeadersStream));
```
Comment (halter73) [inline src/Kestrel.Core/Internal/Http2/Http2Connection.cs]: Nit: For more @pranavkm points you can shorten this to

```C#
ThreadPool.UnsafeQueueUserWorkItem(state =>
{
    var (app, currentStream) = (Tuple<IHttpApplication<TContext>, Http2Stream>)state;
    _ = currentStream.ProcessRequestsAsync(app);
},
(application, _currentHeadersStream));
```

File: src/Kestrel.Core/Internal/Http2/Http2Connection.cs [hunk 3]
```diff
@@ -672,7 +672,15 @@ private void StartStream<TContext>(IHttpApplication<TContext> application)
             }
 
             _streams[_incomingFrame.StreamId] = _currentHeadersStream;
-            _ = _currentHeadersStream.ProcessRequestsAsync(application);
+            // Must not allow app code to block the connection handling loop.
+            System.Threading.ThreadPool.UnsafeQueueUserWorkItem(state =>
```
Comment (halter73) [inline src/Kestrel.Core/Internal/Http2/Http2Connection.cs]: Nit: Add using.

File: src/Kestrel.Core/Internal/Http2/Http2Connection.cs [hunk 4]
```diff
@@ -672,7 +672,15 @@ private void StartStream<TContext>(IHttpApplication<TContext> application)
...
+                {
```
Comment (halter73) [inline src/Kestrel.Core/Internal/Http2/Http2Connection.cs]: Nit: In Kestrel we consistently use the same level of indentation for the `{` after a `=>`.

File: test/Kestrel.Core.Tests/Http2ConnectionTests.cs [hunk 5]
```diff
@@ -131,8 +131,8 @@ public Http2ConnectionTests()
         {
             var inlineSchedulingPipeOptions = new PipeOptions(
                 pool: _memoryPool,
-                readerScheduler: PipeScheduler.Inline,
-                writerScheduler: PipeScheduler.Inline,
+                readerScheduler: PipeScheduler.ThreadPool,
+                writerScheduler: PipeScheduler.ThreadPool,
```
Comment (halter73) [inline test/Kestrel.Core.Tests/Http2ConnectionTests.cs]: This is probably responsible for the ConnectionResetLoggedWithActiveStreams failure. I don't think you should change this. If for some reason it's needed for a new test, use `threadPoolSchedulingPipeOptions` for just the new test.

File: test/Kestrel.Core.Tests/Http2ConnectionTests.cs [hunk 6]
```diff
@@ -641,7 +707,15 @@ void IHttpHeadersHandler.OnHeader(Span<byte> name, Span<byte> value)
                 ignoreNonGoAwayFrames: false,
                 expectedLastStreamId: 1,
                 expectedErrorCode: Http2ErrorCode.STREAM_CLOSED,
-                expectedErrorMessage: CoreStrings.FormatHttp2ErrorStreamClosed(Http2FrameType.DATA, streamId: 1));
+                expectedErrorMessage: null);
+
+            // There's a race where either of these messages could be logged, depending on if the stream cleanup has finished yet.
```
Comment (halter73) [inline test/Kestrel.Core.Tests/Http2ConnectionTests.cs]: Was this race there before or just after changing the scheduling mode? If it's the latter, revert this test change.

File: src/Kestrel.Core/Internal/Http2/Http2Connection.cs (hunk 2 above)
Comment (gfoidl) [reply src/Kestrel.Core/Internal/Http2/Http2Connection.cs]: With `UnsafeQueueUserWorkItem` the state-argument (valuetuple) gets boxed. This allocation can be avoided with the generic `QueueUserWorkItem`-overload:
```c#
ThreadPool.QueueUserWorkItem(state =>
{
    var (app, currentStream) = state;
    _ = currentStream.ProcessRequestsAsync(app);
},
(application, _currentHeadersStream), false);
```
Or is the "unsafe queuing" needed here?

Further the allocation for the delegate could be avoided by caching it.

Comment (Tratcher) [reply src/Kestrel.Core/Internal/Http2/Http2Connection.cs]: Unsafe is more efficient as it avoids capturing the current thread context.

Comment (Tratcher) [reply src/Kestrel.Core/Internal/Http2/Http2Connection.cs]: I'm not able to fully extract the delegate due to the method's generic type.

Comment (benaadams) [reply src/Kestrel.Core/Internal/Http2/Http2Connection.cs]: > Unsafe is more efficient as it avoids capturing the current thread context.

Can always do this instead for a similar effect
```csharp
bool restoreFlow = false;
try
{
    if (!ExecutionContext.IsFlowSuppressed())
    {
        ExecutionContext.SuppressFlow();
        restoreFlow = true;
    }

    ThreadPool.QueueUserWorkItem(state =>
    {
        var (app, currentStream) = state;
            _ = currentStream.ProcessRequestsAsync(app);
    },
    (application, _currentHeadersStream), false);
}
finally
{
    // Restore the current ExecutionContext
    if (restoreFlow)
        ExecutionContext.RestoreFlow();
}
```
As per https://github.com/dotnet/corefx/pull/26077

Comment (Tratcher) [reply src/Kestrel.Core/Internal/Http2/Http2Connection.cs]: What's the advantage to doing that over UnsafeQueueUserWorkItem?

Comment (benaadams) [reply src/Kestrel.Core/Internal/Http2/Http2Connection.cs]: Depends if you can use the generic state overload (on netcoreapp2.1) to avoid the boxing of the state
```csharp
QueueUserWorkItem<TState>(Action<TState> callBack, TState state, bool preferLocal)
```

Comment (benaadams) [reply src/Kestrel.Core/Internal/Http2/Http2Connection.cs]: Also is an `Action<T>` callback rather than a `WaitCallback`

Comment (Tratcher) [reply src/Kestrel.Core/Internal/Http2/Http2Connection.cs]: We're trying to limit our use of netcoreapp2.1 APIs to only places where we can't avoid it. If this shows up in the profiles then we may try adding a fast-path.

Comment (benaadams) [reply src/Kestrel.Core/Internal/Http2/Http2Connection.cs]: Is no advantage vs `UnsafeQueueUserWorkItem` on netstandard2.0

Comment (gfoidl) [reply src/Kestrel.Core/Internal/Http2/Http2Connection.cs]: Besides "limit our use of netcoreapp2.1 APIs" would it make sense to make `UnsafeQueueUserWorkItem` generic similar to https://github.com/dotnet/coreclr/pull/16570?

Comment (halter73) [reply src/Kestrel.Core/Internal/Http2/Http2Connection.cs]: I was suggesting the new syntax for terseness not for valuetuple. We don't want to ifdef to use netcoreapp2.1 APIs unless theres a functional improvement or significant perf improvement. Explicitly call ExecutionContext.Suppress/RestoreFlow is also pretty gross.

File: src/Kestrel.Core/Internal/Http2/Http2Connection.cs [hunk 7]
```diff
@@ -672,7 +672,13 @@ private void StartStream<TContext>(IHttpApplication<TContext> application)
...
 
             _streams[_incomingFrame.StreamId] = _currentHeadersStream;
-            _ = _currentHeadersStream.ProcessRequestsAsync(application);
+            // Must not allow app code to block the connection handling loop.
+            ThreadPool.UnsafeQueueUserWorkItem(state =>
+            {
+                var (app, currentStream) = (Tuple<IHttpApplication<TContext>, Http2Stream>)state;
+                _ = currentStream.ProcessRequestsAsync(app);
+            },
+            new Tuple<IHttpApplication<TContext>,Http2Stream>(application, _currentHeadersStream));
```
Comment (halter73) [inline src/Kestrel.Core/Internal/Http2/Http2Connection.cs]: Nit: missing space after `>,`.

Also if we really want to avoid an allocation here, we could just add a reference to the application in Http2Stream.

**Comment by halter73** [comment]
It looks like dispatching might be causing some [flakiness](https://dotnet.visualstudio.com/public/public%20Team/_build/results?buildId=2362&view=logs) with Http2ConnectionTests.HEADERS_Received_StreamClosed_ConnectionError.

```
2018-06-18T19:36:23.5227430Z   [xUnit.net 00:00:05.9252730]     HEADERS_Received_StreamClosed_ConnectionError [FAIL]
2018-06-18T19:36:23.5244610Z   Failed   HEADERS_Received_StreamClosed_ConnectionError
2018-06-18T19:36:23.5368000Z RUNDOTNET : error Message:  [/Users/vsts/agent/2.134.2/work/.dotnet/buildtools/korebuild/2.2.0-preview1-17087/KoreBuild.proj]
2018-06-18T19:36:23.5385250Z    Assert.Contains() Failure
2018-06-18T19:36:23.5400530Z   Not found: The client sent a HEADERS frame to closed stream ID 1.
2018-06-18T19:36:23.5418860Z In value : HTTP/2 connection error (STREAM_CLOSED): The client sent a HEADERS frame to stream ID 1 which is in the "half-closed (remote) state". [/Users/vsts/agent/2.134.2/work/.dotnet/buildtools/korebuild/2.2.0-preview1-17087/KoreBuild.proj]
2018-06-18T19:36:23.5436600Z   Stack Trace:
2018-06-18T19:36:23.5452880Z      at Microsoft.AspNetCore.Server.Kestrel.Core.Tests.Http2ConnectionTests.WaitForConnectionErrorAsync[TException](Boolean ignoreNonGoAwayFrames, Int32 expectedLastStreamId, Http2ErrorCode expectedErrorCode, String expectedErrorMessage)
2018-06-18T19:36:23.5469770Z      at Microsoft.AspNetCore.Server.Kestrel.Core.Tests.Http2ConnectionTests.HEADERS_Received_StreamClosed_ConnectionError() in /_/test/Kestrel.Core.Tests/Http2ConnectionTests.cs:line 1014
2018-06-18T19:36:23.5487510Z   --- End of stack trace from previous location where exception was thrown ---
```

File: src/Kestrel.Core/Internal/Http2/Http2Connection.cs (hunk 7 above)
Comment (Tratcher) [reply src/Kestrel.Core/Internal/Http2/Http2Connection.cs]: That generic is very viral. Right now it's only passed through method signatures so it can't be stored in any member fields.

Comment (halter73) [reply src/Kestrel.Core/Internal/Http2/Http2Connection.cs]: You could always cast from object in the callback, but I wouldn't bother.

Comment (gfoidl) [reply src/Kestrel.Core/Internal/Http2/Http2Connection.cs]: > Nit: missing space after `>,`.

Reminder :wink:
